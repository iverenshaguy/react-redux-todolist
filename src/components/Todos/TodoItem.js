import React, { Component } from 'react';
import PropTypes from 'prop-types';
import AddTodo from './AddTodo';
import EditTodo from './EditTodo';
import DeleteTodo from './DeleteTodo';
import TodoStatus from './TodoStatus';
import './TodoItem.css';

class TodoItem extends Component {
  constructor() {
    super();
    this.state = {
      todoTitle: 'show',
      editForm: 'hide'
    };
  }

  editTodoToggle = () => {
    if (this.state.editForm === 'hide') {
      this.setState({
        todoTitle: 'hide',
        editForm: 'show'
      });
    } else {
      this.setState({
        todoTitle: 'show',
        editForm: 'hide'
      });
    }
  };

  handleEditTodo = todo => {
    const updateTodo = new Promise((resolve, reject) => {
      this.props.onEdit(todo);
      resolve();
    });

    updateTodo.then(() => {
      this.setState({
        todoTitle: 'show',
        editForm: 'hide'
      });
    });
  };

  handleTodoStatus = () => {
    const { todo } = this.props;
    const newTodo = {
      userId: 1,
      id: todo.id,
      title: todo.title,
      completed: !todo.completed
    };

    this.props.onEdit(newTodo);
  };

  handleDeleteTodo = () => {
    const id = this.props.todo.id;
    this.props.onDelete(id);
  };

  render() {
    const { editForm, todoTitle } = this.state;
    const { todo } = this.props;

    let completed;

    if (todo.completed) {
      completed = 'Completed';
    } else {
      completed = 'Not Completed';
    }

    return (
      <div className="TodoItem">
        <div className="card border-primary">
          <div className="card-body">
            <DeleteTodo onClick={this.handleDeleteTodo} />
            <blockquote className="card-blockquote">
              <p className={todoTitle}>
                {todo.title}&nbsp;&nbsp;
                <span className="badge badge-primary">{completed}</span>
              </p>
              <div className={editForm}>
                <AddTodo
                  todo={todo}
                  buttonText="Update Todo"
                  editTodo={this.handleEditTodo}
                />
              </div>
              <br />
              <footer>
                <EditTodo onClick={this.editTodoToggle} />&nbsp;&nbsp;
                <TodoStatus todo={todo} onClick={this.handleTodoStatus} />
              </footer>
            </blockquote>
          </div>
        </div>
        <br />
      </div>
    );
  }
}

TodoItem.proptypes = {
  todo: PropTypes.object.isRequired,
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default TodoItem;
