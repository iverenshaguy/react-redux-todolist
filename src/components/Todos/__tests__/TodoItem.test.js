import TodoItem from '../TodoItem';
import { todos as mockTodos } from '../../../__mocks__/todos.mock';

const deleteTodoFunction = jest.fn();
const editTodoFunction = jest.fn();
const setup = () => {
  const props = {
    todo: mockTodos[0],
    onEdit: editTodoFunction,
    onDelete: deleteTodoFunction
  };

  const shallowWrapper = shallow(<TodoItem {...props} />);
  const mountedWrapper = mount(<TodoItem {...props} />);

  return {
    shallowWrapper,
    mountedWrapper,
    props
  };
};

describe('TodoItem', () => {
  it('renders correctly', () => {
    const { shallowWrapper } = setup();
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });

  it('mounts correctly', () => {
    const { mountedWrapper } = setup();
    expect(toJson(mountedWrapper)).toMatchSnapshot();
  });

  it('toggles todo form and todo title with edit todo click', () => {
    const { mountedWrapper } = setup();
    mountedWrapper.setState({
      editForm: 'hide'
    });
    mountedWrapper.find('EditTodo').simulate('click');
    expect(mountedWrapper.instance().state.todoTitle).toEqual('hide');
    expect(mountedWrapper.instance().state.editForm).toEqual('show');

    mountedWrapper.find('EditTodo').simulate('click');
    expect(mountedWrapper.instance().state.todoTitle).toEqual('show');
    expect(mountedWrapper.instance().state.editForm).toEqual('hide');
  });

  it('changes todoItem status', () => {
    const { mountedWrapper } = setup();
    const submitedTodo = {
      userId: 1,
      id: 1,
      title: ' Levies and I ',
      completed: false
    };

    mountedWrapper.find('TodoStatus').simulate('click');
    expect(editTodoFunction).toHaveBeenCalledWith(submitedTodo);
  });

  it('edits a todoItem and hides form', () => {
    const { mountedWrapper } = setup();
    const event = { target: { value: 'Learn React Test Driven Development' } };
    const submitedTodo = {
      userId: 1,
      id: 1,
      title: 'Learn React Test Driven Development',
      completed: false
    };
    mountedWrapper.find('input').simulate('change', event);
    mountedWrapper.find('form').simulate('submit');

    expect(editTodoFunction).toHaveBeenCalledWith(submitedTodo);
    expect(mountedWrapper.instance().state.todoTitle).toEqual('show');
    expect(mountedWrapper.instance().state.editForm).toEqual('hide');
  });

  it('deletes a todoItem', () => {
    const { mountedWrapper } = setup();
    mountedWrapper.find('DeleteTodo').simulate('click');
    expect(deleteTodoFunction).toHaveBeenCalledWith(mockTodos[0].id);
  });

  it("shows 'Completed' according to passed in prop", () => {
    const { shallowWrapper } = setup();
    expect(shallowWrapper.find('.badge-primary').html()).toContain('Completed');
  });

  it("shows 'Unompleted' according to passed in prop", () => {
    const wrapper = shallow(
      <TodoItem {...setup().props} todo={mockTodos[2]} />
    );
    expect(wrapper.find('.badge-primary').html()).toContain('Not Completed');
  });
});
