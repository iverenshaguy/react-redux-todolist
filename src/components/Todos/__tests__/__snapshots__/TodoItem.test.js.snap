// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoItem mounts correctly 1`] = `
<TodoItem
  onDelete={[Function]}
  onEdit={[Function]}
  todo={
    Object {
      "completed": true,
      "id": 1,
      "title": " Levies and I ",
    }
  }
>
  <div
    className="TodoItem"
  >
    <div
      className="card border-primary"
    >
      <div
        className="card-body"
      >
        <DeleteTodo
          onClick={[Function]}
        >
          <Button
            btnClass="close"
            buttonText="×"
            buttonType="button"
            onClick={[Function]}
          >
            <button
              className="close"
              onClick={[Function]}
              type="button"
            >
              ×
            </button>
          </Button>
        </DeleteTodo>
        <blockquote
          className="card-blockquote"
        >
          <p
            className="show"
          >
             Levies and I 
              
            <span
              className="badge badge-primary"
            >
              Completed
            </span>
          </p>
          <div
            className="hide"
          >
            <AddTodo
              buttonText="Update Todo"
              editTodo={[Function]}
              todo={
                Object {
                  "completed": true,
                  "id": 1,
                  "title": " Levies and I ",
                }
              }
            >
              <div
                className="AddTodo"
              >
                <form
                  className="form-inline my-2 my-lg-0"
                  onSubmit={[Function]}
                >
                  <input
                    autoFocus={true}
                    className="form-control mr-sm-2 new-todo"
                    name="todoinput"
                    onChange={[Function]}
                    placeholder="Enter Todo"
                    type="text"
                    value=""
                  />
                  <button
                    className="btn btn-primary my-2 my-sm-0"
                    type="submit"
                  >
                    Update Todo
                  </button>
                </form>
              </div>
            </AddTodo>
          </div>
          <br />
          <footer>
            <EditTodo
              onClick={[Function]}
            >
              <Button
                btnClass="btn btn-outline-primary"
                buttonText="Edit Todo"
                buttonType="button"
                onClick={[Function]}
              >
                <button
                  className="btn btn-outline-primary"
                  onClick={[Function]}
                  type="button"
                >
                  Edit Todo
                </button>
              </Button>
            </EditTodo>
              
            <TodoStatus
              onClick={[Function]}
              todo={
                Object {
                  "completed": true,
                  "id": 1,
                  "title": " Levies and I ",
                }
              }
            >
              <Button
                btnClass="btn btn-outline-primary"
                buttonText="Mark as Undone"
                buttonType="button"
                onClick={[Function]}
              >
                <button
                  className="btn btn-outline-primary"
                  onClick={[Function]}
                  type="button"
                >
                  Mark as Undone
                </button>
              </Button>
            </TodoStatus>
          </footer>
        </blockquote>
      </div>
    </div>
    <br />
  </div>
</TodoItem>
`;

exports[`TodoItem renders correctly 1`] = `
<div
  className="TodoItem"
>
  <div
    className="card border-primary"
  >
    <div
      className="card-body"
    >
      <DeleteTodo
        onClick={[Function]}
      />
      <blockquote
        className="card-blockquote"
      >
        <p
          className="show"
        >
           Levies and I 
            
          <span
            className="badge badge-primary"
          >
            Completed
          </span>
        </p>
        <div
          className="hide"
        >
          <AddTodo
            buttonText="Update Todo"
            editTodo={[Function]}
            todo={
              Object {
                "completed": true,
                "id": 1,
                "title": " Levies and I ",
              }
            }
          />
        </div>
        <br />
        <footer>
          <EditTodo
            onClick={[Function]}
          />
            
          <TodoStatus
            onClick={[Function]}
            todo={
              Object {
                "completed": true,
                "id": 1,
                "title": " Levies and I ",
              }
            }
          />
        </footer>
      </blockquote>
    </div>
  </div>
  <br />
</div>
`;
